### PR TITLE
Support cross-version upgrades

### LINKED ISSUE #230: automate config file upgrades from previous versions
After the next release (4.3?), future versions of Warewulf should upgrade previous versions (4.2 and earlier excluded).

This would entail migrating configuration of the previous version to the new version, retaining as much as possible:
 - Node and server configuration files
 - Directory layout
 - Node image, overlay, and container formats

Specifically, the request is to add successful upgrades to the validation requirements for a new release.

### LINKED ISSUE #517: package and document update_configuration
I recently discovered the `update_configuration` utility in the Warewulf source. I think this would be generally useful, particularly as more people start using Warewulf 4 when it gets published in OHPC. With that, I suggest that the RPM spec file be updated to include the update_configuration command.

It could be distributed as-is in a non-PATH location (e.g., something like a libexec); but if we put it in `*bin/` I suggest it could be renamed to something like `wwcfg`.

### LINKED ISSUE #1507: remove WW_INTERNAL in nodes.conf and warewulf.conf
I don't think this is actually used. We should just eliminate it, and then make legacy readers that can import content from previous versions formats.

### LINKED ISSUE #1508: rename and refactor nodes datastructures
- NodesConf
- WarewulfConf
- Node
- Profile
- NetDev
- Tag
- NetDevTag

### PR DESCRIPTION
**Pull request by anderbubble**
## Description of the Pull Request (PR):

Add a new `wwctl upgrade` to automatically convert `warewulf.conf` and `nodes.conf` to the current format.

Also includes a minor cleanup in the names of the node structs.

### todo

- [x] add `wwctl upgrade` subcommand
  - [x] defaults boolean
  - [x] source file
  - [x] output file
  - [x] generic -> split
  - [x] wwinit -> split
- [x] add default values now that they are missing
- [x] complete the test suite
  - [x] mountoptions interface
- [x] warewulf.conf upgrades
- [x] documentation

### defaults

```
defaultnode:
  runtime overlay:
  - hosts
  - ssh.authorized_keys
  - syncuser
  system overlay:
  - wwinit
  - wwclient
  - fstab
  - hostname
  - ssh.host_keys
  - issue
  - resolv
  - udev.netname
  - systemd.netname
  - ifcfg
  - NetworkManager
  - debian.interfaces
  - wicked
  - ignition
  kernel:
    args: quiet crashkernel=no vga=791 net.naming-scheme=v238
  init: /sbin/init
  root: initramfs
  ipxe template: default
  profiles:
  - default
  network devices:
    dummy:
      type: ethernet
      netmask: 255.255.255.0
```

## This fixes or addresses the following GitHub issues:

- Closes #230 
- Closes #517
- Closes #1507 
- Closes #1508

## Reviewer  checklist

The reviewer checks the following items before merging the PR.

- [ ] The PR is based on the appropriate branch (typically [main](https://github.com/warewulf/warewulf/tree/main/userdocs))
- [ ] All commits are "Signed off" (e.g., using `git commit --signoff`) in agreement to the [DCO](DCO.txt)
- [ ] The [CHANGELOG](https://github.com/warewulf/warewulf/blob/main/CHANGELOG.md) has been updated, if necessary, and under the correct release heading
- [ ] The [userdocs](https://github.com/warewulf/warewulf/tree/main/userdocs) have been updated, if necessary
- [ ] The submitter is listed in the [contributors file](https://github.com/warewulf/warewulf/blob/main/CONTRIBUTORS.md)
- [ ] The test suite has been updated, if necessary

### CODE REVIEW

**Comment by anderbubble** [comment]
I found a bug in the `warewulf.conf` handling, so I'm making a draft again for now.

**Comment by mslacken** [comment]
Thanks for your great work. 
I have following points:
* I still would keep the `WW_INTERNAL` so that it's possible to know which version we must parse
* create a new `wwctl config` command, which will then have the sub commands `upgrade`, `genconf` and `clean` command

**Comment by anderbubble** [comment]
> we have Colorado based contributor

I expected you'd give me a hard time about that. ^_^

fwiw, the "refactoring" is just a simple rename of some of the entities, I think to make them clearer-to-purpose. Any other changes that have been made were only for a functional purpose in support of upgradability.

**Comment by anderbubble** [comment]
> create a new wwctl config command, which will then have the sub commands upgrade, genconf and clean command

@mslacken how would you feel about merging this as-is, and then reorganizing the commands as you've proposed as a separate PR? Since it'd affect multiple otherwise-unrelated commands, I think that's neater, rather than continuing to broaden this PR.

**Comment by anderbubble** [comment]
> * I still would keep the `WW_INTERNAL` so that it's possible to know which version we must parse

As discussed in the TSC meeting, the `WW_INTERNAL` field isn't necessary as the upgrade code can always just do the best it can with the fields it has, and knowing the version wouldn't change that.

**Comment by anderbubble** [comment]
A note to reviewers:

You will see many changes like this: https://github.com/warewulf/warewulf/pull/1524/files#diff-5eeb0c491c481c4618fc1da56ca39612a315603ee94dc26f7a5239f5a74aba8fL105-R105

This is because booleans in the warewulf.conf struct have been converted to `*bool`. This is so that `false` is considered non-empty, and `nil` is considered empty. In other words, this allows us to distinguish between "not specified," "true," and "false."

But text/template considers any non-nil pointer to be truthy, and doesn't support dereferencing `*bool`s. To alleviate this, I renamed all such struct fields from `Name` to `NameP` (`P` for "pointer") and defined a matching `Name() bool` method. text/template runs such methods automatically, so `struct.Name` still works in templates unmodified.

So it's a lot of small change, but I think it's all on the backend in support of an unmodified user experience.

**Comment by anderbubble** [comment]
rebased to main

**Comment by anderbubble** [comment]
@mslacken I rebased onto current main since there have been a few consequential merges. Let me know what needs to be done to get this PR approved.

File: internal/pkg/config/buildconfig.go.in [hunk 1]
```diff
@@ -24,28 +26,49 @@ type BuildConfig struct {
...
 const Release = "@RELEASE@"
 
-// must be set .in file so that its available for tests
-const Confversion = "45"
-
 type TFTPConf struct {
-	Enabled     bool   `yaml:"enabled" default:"true"`
-	TftpRoot    string `yaml:"tftproot" default:"@TFTPDIR@"`
-	SystemdName string `yaml:"systemd name" default:"tftp"`
+	EnabledP    *bool  `yaml:"enabled" default:"true"`
```
Comment (mslacken) [inline internal/pkg/config/buildconfig.go.in:30]: `EnabledP` <- is the P a typo?

**Review by mslacken — APPROVED** [review]

**Pull request merged**